### DISCUSSION TITLE
Bug: Refresh Token is called when registering a user (create a user) after logout is called

### DISCUSSION THREAD

**Discussion by kent-leow** [directus/directus#13825] [Archive]
### Describe the Bug

Error: "refresh_token" is required in either the JSON payload or Cookie

This error occurred when I am trying to create a new user to Directus right after I called logout.

I am using Directus JS SDK.

### To Reproduce

1. Install newest directus js sdk in any js project.
2. Initiate direct = new Directus(url, {
        auth: {
            mode: 'json',
        },
    });
3. Call await directus.auth.logout()
4. Call await directus.users.createOne({..this is user object..})
5. Error occurs.

### Errors Shown

Error: "refresh_token" is required in either the JSON payload or Cookie

### What version of Directus are you using?

Community cloud

### What version of Node.js are you using?

-

### What database are you using?

-

### What browser are you using?

Chrome

### How are you deploying Directus?

Community cloud

**Comment by br41nslug** [comment]
By default creating users requires admin permissions so this error is to be expected *after* logout. Have you set the right public permissions for creating a public registration like you're describing?

**Reply by kent-leow** [reply]
I've set the access to allow public to create user (as register). I'm not sure why after logout and I try to create a user, it actually trigger refresh token, then error occured.

**Reply by kent-leow** [reply]
@br41nslug is the behaviour same at your side?

**Reply by br41nslug** [reply]
I've not yet been able to reproduce it on my test setup unfortunately if i can i'll re-open the ticket (mainly due to other blocks like perms & cors tbh). 
would you mind sharing the permissions as you've set them up?

**Reply by kent-leow** [reply]
@br41nslug I believe if you set the permission to all allow, the issue will persist. 
I set the permission as below.
<img width="804" alt="image" src="https://user-images.githubusercontent.com/105830954/173806503-d681dac6-4378-4c3f-869f-2a68838ab9d7.png">

Thanks!!

**Reply by kent-leow** [reply]
I used Directus with 'json' instead of 'cookie' (default)

**Reply by br41nslug** [reply]
I have been able to reproduce it now altough on my side it errors on the `await directus.auth.logout()` call not when creating the user. Can you verify this is the case too?

it's not a great fix but checking for a refresh token before logging out fixes it for me:
```js
if (!!directus.storage.get('auth_refresh_token')) {
        await directus.auth.logout();
    }
    await directus.users.createOne({....});
```
 While i think logout should not necessarily throw an error when not logged in it is also not completely preventable. When using json you can check if its set but thats not possible for cookie mode.

**Reply by kent-leow** [reply]
@br41nslug Thank you so much for spending time to reproduce the issue. It will throw error on `await directus.auth.logout()` if you're not calling `await directus.auth.login()` beforehand which I didn't mention in the steps. If you call logout directly before login, it will throw error for sure because you do not have access/refresh token yet.

In my case, I did login (sorry I didn't mention it earlier), then proceed to logout, it should clear the token in storage, somehow, when I proceed to create user (or other collection item), it will call to refresh token when there's no token already, hence error occured.

Need your help to execute this with one more step (no 3). This is probably a bug.

1. Install newest directus js sdk in any js project.
2. Initiate direct = new Directus(url, {
auth: {
mode: 'json',
},
});
3. Call await directus.auth.login({email and password here})
4. Call await directus.auth.logout()
5. Call await directus.users.createOne({..this is user object..})
6. Error occurs.


Thank you so much :)

**Reply by br41nslug** [reply]
that makes sense, but if i add the login first im not getting the error unfortunately.
updated my test code to:
```js
    const directus = new Directus(process.env.REACT_APP_DIRECTUS_URL, {
        auth: {
            mode: 'json',
        },
    });
    console.log('login', await directus.auth.login({ email: 'admin@example.com', password: 'd1r3ctu5' }));
    console.log('logout', await directus.auth.logout());
    console.log('register User', await directus.users.createOne({ email: 'register123@example.com', password: 'test' }));
```
And that results the following successful logs (tested in chrome and firefox):
```
login {access_token: '...token...', refresh_token: '...refresh...', expires: 900000}
logout undefined
register User {id: '345e1a37-ecee-4d5d-ad68-1d4228965e61', first_name: null, last_name: null, email: 'register123@example.com', password: '**********', …}
```

**Reply by kent-leow** [reply]
@br41nslug Thanks! It's weird. My result:

![image](https://user-images.githubusercontent.com/105830954/173873476-ad85c977-8c06-42bd-bdb6-bc57c738477a.png)

Not sure what's the issue.

**Reply by br41nslug** [reply]
In that screenshot it seems to be errorring on a call to `/auth/refresh` i'm not seeing that refresh call happening on my side for some reason :(

**Reply by kent-leow** [reply]
Hi @br41nslug , thanks for reply. May I request you to test on next.js project, it seems that issue only happens on next.js. Thanks!

**Reply by br41nslug** [reply]
Interesting you are very right. For some reason this does occur specifically in the next.js run time 🤔 Must be some of the SSG/SSR features interfering

**Reply by MrCasquette** [reply]
I can confirm. I had the same "bug". But actually I think a few person only will login and register after 😅 However it's not an expected behavior. BTW Do we should handle login/logout in getStaticProps or with a simple useEffect?

**Reply by MrCasquette** [reply]
@br41nslug sorry for the tag but have you find any solution to this specific problem? I noticed than when i refresh the page and then try to logout the same error appears whereas when I just logout without refreshing before, it works perfectly.
